core/src/main/java/org/elasticsearch/rest/BytesRestResponse.java

```diff
@@ -116,31 +119,23 @@ public RestStatus status() {
 
     private static final Logger SUPPRESSED_ERROR_LOGGER = ESLoggerFactory.getLogger("rest.suppressed");
 
-    private static XContentBuilder convert(RestChannel channel, RestStatus status, Exception e) throws IOException {
-        XContentBuilder builder = channel.newErrorBuilder().startObject();
-        if (e == null) {
-            builder.field("error", "unknown");
-        } else if (channel.detailedErrorsEnabled()) {
-            final ToXContent.Params params;
-            if (channel.request().paramAsBoolean("error_trace", !ElasticsearchException.REST_EXCEPTION_SKIP_STACK_TRACE_DEFAULT)) {
-                params =  new ToXContent.DelegatingMapParams(
-                        Collections.singletonMap(ElasticsearchException.REST_EXCEPTION_SKIP_STACK_TRACE, "false"), channel.request());
+    private static XContentBuilder build(RestChannel channel, RestStatus status, Exception e) throws IOException {
+        ToXContent.Params params = channel.request();
+        if (params.paramAsBoolean("error_trace", !REST_EXCEPTION_SKIP_STACK_TRACE_DEFAULT)) {
+            params =  new ToXContent.DelegatingMapParams(singletonMap(REST_EXCEPTION_SKIP_STACK_TRACE, "false"), params);
+        } else if (e != null) {
+            Supplier<?> messageSupplier = () -> new ParameterizedMessage("path: {}, params: {}",
+                    channel.request().rawPath(), channel.request().params());
+
+            if (status.getStatus() < 500) {
+                SUPPRESSED_ERROR_LOGGER.debug(messageSupplier, e);
             } else {
-                if (status.getStatus() < 500) {
-                    SUPPRESSED_ERROR_LOGGER.debug(
-                            (Supplier<?>) () -> new ParameterizedMessage("path: {}, params: {}",
-                                    channel.request().rawPath(), channel.request().params()), e);
-                } else {
-                    SUPPRESSED_ERROR_LOGGER.warn(
-                            (Supplier<?>) () -> new ParameterizedMessage("path: {}, params: {}",
-                                    channel.request().rawPath(), channel.request().params()), e);
-                }
-                params = channel.request();
+                SUPPRESSED_ERROR_LOGGER.warn(messageSupplier, e);
             }
-            ElasticsearchException.renderException(builder, params, e);
-        } else {
-            builder.field("error", simpleMessage(e));
         }
+
+        XContentBuilder builder = channel.newErrorBuilder().startObject();
+        ElasticsearchException.generateFailureXContent(builder, params, e, channel.detailedErrorsEnabled());
         builder.field("status", status.getStatus());
         builder.endObject();
```

Comment:
it's unfortunate that we still have the status here, it means that when writing parsing code `failureFromXContent` and `throwableFromXContent` won't be enough, we will still need a parse method as part of this class.

I wonder if we can remove the `public BytesRestResponse(RestChannel channel, RestStatus status, Exception e) throws IOException {` constructor, then the status always comes from the exception and this bit can be moved out too?